### PR TITLE
Use perspective API to detect toxicity of comments

### PR DESCRIPTION
**Pull request by swetha-gangu**
- pulled in java perspective API and put it in it's own package 
- called the API before adding a comment in datastore and only added if it is not toxic
- need to add functionality to direct the user to add a less toxic comment 
- need to handle exceptions

### CODE REVIEW

File: src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
+import java.lang.IllegalStateException; 
+import java.util.Map;
 import java.util.Objects;
 import com.google.sps.data.Constants;
 
 public final class Comment {
     private final String name;
     private final String comment;
     private final long id;
+    private static final String apiKey = System.getenv(Constants.PERSPECTIVE_API_KEY);
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:16]: put static variables above instance variabels

File: src/main/java/com/google/sps/data/Comment.java [hunk 2]
```diff
@@ -26,6 +33,21 @@ public long getID() {
         return id; 
     }
 
+    public static boolean isCommentToxic(String msg) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:36]: public static methods should go above the constructor

go/java-practices/class-structure.md

File: src/main/java/com/google/sps/data/Comment.java [hunk 3]
```diff
@@ -26,6 +33,21 @@ public long getID() {
...
+        PerspectiveAPI pAPI = PerspectiveAPI.create(apiKey); 
+        AnalyzeCommentResponse response = pAPI.analyze(msg); 
+        if (response == null) {
+            throw new IllegalStateException("Perspective API was unable to analyze your comment"); 
+        }
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:41]: add a newline after this block

File: src/main/java/com/google/sps/data/Comment.java [hunk 4]
```diff
@@ -26,6 +33,21 @@ public long getID() {
...
+        Map<AttributeType, AttributeScore> scores = response.getAttributeScores(); 
+        if (!scores.containsKey(AttributeType.TOXICITY)) {
+            return false; 
+        }
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:45]: add a newline after this block

File: src/main/java/com/google/sps/data/Comment.java (hunk 2 above)
Comment (v1shanker) [inline src/main/java/com/google/sps/data/Comment.java:36]: this method should either go in a new file ToxicityDetector.java (or something like that) and take in a String. or this method should be an instance method and just operator on this.comment.

If you have questions about this, feel free to ask.

File: src/main/java/com/google/sps/data/ToxicityDetector.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        if (!scores.containsKey(AttributeType.TOXICITY)) {
+            return false; 
+        }
+        
+        AttributeScore score = scores.get(AttributeType.TOXICITY);
+        double toxicity = score.getSummaryScore().getValue(); 
+        return toxicity >= Constants.COMMENT_TOXICITY_THRESHOLD; 
+    }
+
+    public ToxicityDetector() {
```
Comment (epoms) [inline src/main/java/com/google/sps/data/ToxicityDetector.java:31]: constructors go before regular methods

Comment (v1shanker) [reply src/main/java/com/google/sps/data/ToxicityDetector.java:31]: constructor can be private

File: src/main/java/com/google/sps/data/DatastoreManager.java [hunk 6]
```diff
@@ -34,6 +36,10 @@
      * @return ID of entity inserted into datastore
      */ 
     public static long insertCommentInDatastore(String name, String message) {
+        ToxicityDetector detector = new ToxicityDetector(); 
+        if (detector.isCommentToxic(message)) {
+            throw new IllegalArgumentException("Can't enter toxic comments.");
```
Comment (epoms) [inline src/main/java/com/google/sps/data/DatastoreManager.java:41]: I wouldn't use an IllegalArgumentException. For two main reasons:
1. Its stretching the definition of the exception, which is generally used for obviously wrong input. Say an empty string or null object, possibly indicating a bug somewhere else in the code. Whether a comment is toxic or not isn't obvious and could happen all the time.
2. Its a runtime exception (i.e unchecked exception) which is meant for unexpected behavior. 

Here toxic comments are expected to happen and you want any caller of this method to know and handle that accordingly. So you want a checked exception that enforces that. You can create your own exception type such as ToxicCommentException or more generically a DatastoreException that you could reuse across this class.

Happy to explain more.

File: src/main/java/com/google/sps/data/ToxicityDetector.java (hunk 5 above)
Comment (swetha-gangu) [reply src/main/java/com/google/sps/data/ToxicityDetector.java:31]: shouldn't a private constructor go at the end

Comment (v1shanker) [reply src/main/java/com/google/sps/data/ToxicityDetector.java:31]: yes private constructor is at the bottom. I think Eric's comment was before you made it private

Comment (epoms) [reply src/main/java/com/google/sps/data/ToxicityDetector.java:31]: it was

File: src/main/java/com/google/sps/data/DatastoreManager.java (hunk 6 above)
Comment (v1shanker) [reply src/main/java/com/google/sps/data/DatastoreManager.java:40]: +1 please address this.

File: src/main/java/com/google/sps/data/ToxicityDetector.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Map;
+import com.google.sps.data.Constants;
+
+public final class ToxicityDetector {
+    private static final String apiKey = System.getenv(Constants.PERSPECTIVE_API_KEY); 
+
+    public static boolean isCommentToxic(String msg) {
+        PerspectiveAPI pAPI = PerspectiveAPI.create(apiKey); 
+        AnalyzeCommentResponse response = pAPI.analyze(msg); 
+        if (response == null) {
```
Comment (v1shanker) [inline src/main/java/com/google/sps/data/ToxicityDetector.java:17]: we discussed failing open. please use that behavior.

**Pull request merged**